app/trpc/routes/users.ts

```diff
@@ -84,4 +90,229 @@ export const trpcRoutesUsers = {
     .mutation(async ({ input, ctx }) => {
       await db.update(T.users).set(input).where(E.eq(T.users.id, ctx.user.id));
     }),
+
+  users_requestUpdateEmail: procedureBuilder
+    .use(middlewares.requireUser)
+    .input(
+      z.object({
+        email: z.string().email(),
+      })
+    )
+    .mutation(async ({ input, ctx }) => {
+      // we only minimally check if new email is differrnt from current email.
+      // we don't check whether new email already exists in db
+      // because showing an error for that would give away other user's email.
+      tinyassert(input.email !== ctx.user.email);
+      const code = await generateUniqueCode((code) =>
+        selectOne(
+          T.emailUpdateRequests,
+          E.eq(T.emailUpdateRequests.code, code)
+        ).then((row) => !row)
+      );
+      await db.insert(T.emailUpdateRequests).values({
+        userId: ctx.user.id,
+        email: input.email,
+        code,
+      });
+      await sendEmailChangeVerificationEmail({
+        username: ctx.user.username,
+        email: input.email,
+        code,
+      });
+    }),
+
+  users_requestResetPassword: procedureBuilder
+    .input(
+      z.object({
+        email: z.string().email(),
+      })
+    )
+    .mutation(async ({ input }) => {
+      const code = await generateUniqueCode((code) =>
+        selectOne(
+          T.passwordResetRequests,
+          E.eq(T.passwordResetRequests.code, code)
+        ).then((row) => !row)
+      );
+      await db.insert(T.passwordResetRequests).values({
+        email: input.email,
+        code,
+      });
+
+      const user = await selectOne(T.users, E.eq(T.users.email, input.email));
+      if (user) {
+        // hang promise for consistent response time regardless of email validity
+        sendResetPasswordEmail({ email: input.email, code });
+      }
```

Comment:
This might now work as intended on aws lambda as runtime can shutdown before it finishes?

Reply:
Fixed in https://github.com/hi-ogawa/ytsub-v3/pull/401/commits/2e31ddcf4d581949f136d9eacb80606ca0b97442
For now, don't worry about response time thing...
I'll put captcha instead just in case.